parsers/IN_GJ.py

```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python3
+import re
+import string
+
+import requests
+from arrow import get
+from requests import Session
+from parsers import countrycode
+from parsers.lib import web
+from parsers.lib import IN
+from operator import itemgetter
+
+station_map = {
+    "hydro": ["Ukai  (Hydro)", "Kadana (Hydro)", "SSP(RBPH)"],
+    "thermal": ["Ukai(3-5)+Ukai6", "Wanakbori", "Gandhinagar", "Sikka(3-4)", "KLTPS(1-3)+KLTPS4",
+                "SLPP(I+II)", "Akrimota", "JHANOR"],
+    "wind": ["GMR"],
+    "gas": ["Utran(Gas)(II)", "Dhuvaran (Gas)(I)+(II)", "GIPCL(I)", "GIPCL(II)", "GSEG(I+II)", "GPPC", "CLPI",
+            "KAWAS", "Sugen+ Unosgn"],
+    "nuclear": ["KAPP"],
+    "coal": ["TPAECo", "EPGL(I+II)", "Adani(I+II+III)", "BECL(I+II)", "CGPL"]
+}
+
+
+def fetch_data(country_code):
+    countrycode.assert_country_code(country_code, 'IN-GJ')
+
+    solar_html = web.get_response_soup(country_code, 'https://www.sldcguj.com/RealTimeData/GujSolar.php', session)
+    wind_html = web.get_response_soup(country_code, 'https://www.sldcguj.com/RealTimeData/wind.php', session)
+
+    india_date = get(solar_html.find_all('tr')[0].text.split('\t')[-1].strip(), 'D-MM-YYYY h:m:s')
+
+    solar_value = float(eval(solar_html.find_all('tr')[-1].find_all('td')[-1].text.strip()))
+    wind_value = float(eval(wind_html.find_all('tr')[-1].find_all('td')[-1].text.strip()))
+
+    hydro_value = thermal_value = gas_value = nuclear_value = coal_value = 0.0
+
+    value_map = {
+        "date": india_date.datetime,
+        "solar": solar_value,
+        "hydro": hydro_value,
+        "thermal": thermal_value,
+        "wind": wind_value,
+        "gas": gas_value,
+        "coal": coal_value
+    }
+
+    cookies_params = {
+        'ASPSESSIONIDSUQQQTRD': 'ODMNNHADJFGCMLFFGFEMOGBL',
+        'PHPSESSID': 'a301jk6p1p8d50dduflceeg6l1'
+    }
+
+    # other_html = requests.get('https://www.sldcguj.com/RealTimeData/RealTimeDemand.php', params=cookies_params)
+    rows = web.get_response_soup(country_code, 'https://www.sldcguj.com/RealTimeData/RealTimeDemand.php',
+                                 session).find_all('tr')
+
+    for row in rows:
+        if len(row.find_all('td')) > 3:
+            v1, v2 = (re.sub(r'[\n\t\r]', r'', x.text.strip()) for x in itemgetter(*[0, 3])(row.find_all('td')))
```

Comment:
Can you explain this line? More interested in the why than the how, it works though so not looking for any changes.

Reply:
Only because the html contents are formatted that way. The page is not well formatted and is not consistent for all the tags.

Reply:
You weren't kidding about the html being filthy!